### PR TITLE
pref: performance improvements for vendor dashboard -> order details page -> downloadable product permission section.

### PR DESCRIPTION
**Pull request by nurul-umbhiya**
### All Submissions:

* [ ] My code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)
* [ ] My code satisfies feature requirements
* [ ] My code is tested
* [ ] My code passes the PHPCS tests
* [ ] My code has proper inline documentation
* [ ] I've included related pull request(s) (optional)
* [ ] I've included developer documentation (optional)
* [ ] I've added proper labels to this pull request

### Related Pull Request(s)

* Full PR Link

### Closes 

* Closes https://github.com/getdokan/dokan/issues/2184

### How to test the changes in this Pull Request:

* Steps or issue link

### Changelog entry

*Title*

Detailed Description of the pull request. What was previous behaviour
and what will be changed in this PR.

### Before Changes

Describe the issue before changes with screenshots(s).

### After Changes

Describe the issue after changes with screenshot(s).

### Feature Video (optional)

Link of detailed video if this PR is for a feature.

### PR Self Review Checklist:

* Code is not following code style guidelines
* Bad naming: make sure you would understand your code if you read it a few months from now.
* KISS: Keep it simple, Sweetie (not stupid!).
* DRY: Don't Repeat Yourself.
* Code that is not readable: too many nested 'if's are a bad sign.
* Performance issues
* Complicated constructions that need refactoring or comments: code should almost always be self-explanatory.
* Grammar errors.

### FOR PR REVIEWER ONLY:

> As a reviewer, your feedback should be focused on the idea, not the person. Seek to understand, be respectful, and focus on constructive dialog.

> As a contributor, your responsibility is to learn from suggestions and iterate your pull request should it be needed based on feedback. Seek to collaborate and produce the best possible contribution to the greater whole.

* [ ] Correct — Does the change do what it’s supposed to? ie: code 100% fulfilling the requirements?
* [ ] Secure — Would a nefarious party find some way to exploit this change? ie: everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities?
* [ ] Readable — Will your future self be able to understand this change months down the road?
* [ ] Elegant — Does the change fit aesthetically within the overall style and architecture?

### CODE REVIEW

File: includes/Order/Ajax.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+        $page = 1;
+        if ( ! empty( $_GET['page'] ) ) {
+            $page = absint( $_GET['page'] );
+        }
+
+        $term = isset( $_GET['term'] ) ? (string) wc_clean( wp_unslash( $_GET['term'] ) ) : '';
+        $args = [
+            'status'       => 'publish',
+            'downloadable' => true,
+            'author'       => get_current_user_id(),
```
Comment (shohag121) [inline includes/Order/Ajax.php:53]: We should consider Support for vendor Staff here. Please replace `get_current_user_id()` with `dokan_get_current_user_id()`.

File: includes/Order/Ajax.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+            'page'         => $page,
+            'limit'        => $limit,
+            's'            => $term,
+        ];
+
+        $ids             = wc_get_products( $args );
+        $product_objects = array_filter( array_map( 'wc_get_product', $ids ), 'wc_products_array_filter_readable' );
+
```
Comment (shohag121) [inline includes/Order/Ajax.php:61]: I think mapping product array with `wc_get_product` is unnecessary here. Because we did not include `'return' => 'ids',` thus `wc_get_products( $args );`  will return full Product object.

**Review by shohag121 — CHANGES_REQUESTED** [review]

**Comment by shohag121** [comment]
Please also check PHPCS issues.

File: templates/orders/downloadable.php [hunk 3]
```diff
@@ -5,72 +5,57 @@
...
+                $file = $product->get_file( $download->get_download_id() );
+                // translators: file name.
+                $file_count = isset( $file['name'] ) ? $file['name'] : sprintf( __( 'File %d', 'dokan-lite' ), $file_counter );
+
                 include 'order-download-permission-html.php';
 
-                $loop++;
-                $file_count++;
+                ++$loop;
+                ++$file_counter;
```
Comment (mrabbani) [inline templates/orders/downloadable.php:43]: **Findings:**
- `$file_count` is expected in `order-download-permission-html.php` as a numeric value.
- ` sprintf( __( 'File %d', 'dokan-lite' ), $file_counter )` is misleading the message in `order-download-permission-html.php` in **Line:14-26**

File: includes/Order/Ajax.php [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        if ( ! empty( $_GET['limit'] ) ) {
+            $limit = absint( $_GET['limit'] );
+        }
+
+        $page = 1;
+        if ( ! empty( $_GET['page'] ) ) {
+            $page = absint( $_GET['page'] );
+        }
+
+        $term = isset( $_GET['term'] ) ? (string) wc_clean( wp_unslash( $_GET['term'] ) ) : '';
```
Comment (mrabbani) [inline includes/Order/Ajax.php:49]: It is recommended to use `sanitize_text_field` instead `wc_clean` function.

**Review by mrabbani — CHANGES_REQUESTED** [review]

**Comment by StalinDurjo** [comment]
**Issue: All products are displayed regardless of vendors search query**
**Description**: Vendor has searched for specific product by its exact name, but that product is not highlighted and displayed at top. Instead, all products are being displayed regardless of what vendor searches for.

Please refer to video attached for steps to reproduce
https://jam.dev/c/3d42092c-ad7c-4235-9eca-f2626269dc55

@nurul-umbhiya bhai

**Comment by StalinDurjo** [comment]
**Issue: Vendor is unable to search by product ID**
**Description**: I have checked with develop branch. Vendor was able to search for product by the product ID. But, now it is not working. Vendor cannot search for product by product ID.

Please see video attached for reference
https://jam.dev/c/f6d3e9a4-008c-43b9-b203-da384e60e927

@nurul-umbhiya bhai

**Comment by StalinDurjo** [comment]
**Issue: Downloadable product input box is taking full width**
**Description**: Please see the before after images for "Downloadable Product Permission" input field. The "Grant Access" button is pushed down and the input box is taking full width.

Screenshot 2024-04-04 at 11 21 41 AM

Screenshot 2024-04-04 at 11 22 05 AM

@nurul-umbhiya bhai

**Comment by nurul-umbhiya** [comment]
> **Issue: Downloadable product input box is taking full width** **Description**: Please see the before after images for "Downloadable Product Permission" input field. The "Grant Access" button is pushed down and the input box is taking full width.
> Screenshot 2024-04-04 at 11 21 41 AM Screenshot 2024-04-04 at 11 22 05 AM
> 
> @nurul-umbhiya bhai

Previously the select box was a simple HTML select box, now we include search functionality with select2 library. The search functionality will not facilitate to that small space.

**Comment by StalinDurjo** [comment]
**Issue: Search result blocks downloadable product input field** 
**Description**: If downloadable product input field is at the very bottom of screen, the search result overlaps the input field, making it not possible for vendor to see what they are typing.

Please see the video attached for reference
https://jam.dev/c/1c88ec7e-8f7c-431e-a5af-c84d9c2ebb59

@nurul-umbhiya bhai

**Comment by nurul-umbhiya** [comment]
> **Issue: Search result blocks downloadable product input field** **Description**: If downloadable product input field is at the very bottom of screen, the search result overlaps the input field, making it not possible for vendor to see what they are typing.
> 
> Please see the video attached for reference https://jam.dev/c/1c88ec7e-8f7c-431e-a5af-c84d9c2ebb59
> 
> @nurul-umbhiya bhai

This is library related issue and can't be solve from our end.
